ScalaTest 3.2

cheeseng opened the pull request:
- Updated to use ScalaTest 3.2.13.
- Dropped PhantomJS support.
- Added support for new MS Edge browser support.

mkurz commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ project/project/
 project/target/
 target/
 .idea
+.bsp/
```
Please and new line at end of file.

cheeseng replied:
Changes already made.

mkurz commented on build.sbt:
```diff
@@ -102,17 +101,23 @@ lazy val `scalatestplus-play` = project
     organization := "org.scalatestplus.play",
     libraryDependencies ++= Seq(
+      // Note: It seems like the only jackson version that works with all play, play-ws, play-json and selenium is 2.11.4.
+      //       play will bring in older version of selenium as well, so need excluding and let scalatest+selenium brings in the newer version.4
```
If I remember correctly, this is because new selenium versions do not support Java 8 anymore, but right now Play still is build with Java 8.
However we are going to drop Java 8 very soon, so that would resolve that problem.

mkurz commented on project/plugins.sbt:
```diff
@@ -13,3 +13,5 @@ addSbtPlugin("org.scalameta" % "sbt-scalafmt"    % "2.4.6")
 
 addSbtPlugin("com.dwijnand" % "sbt-dynver" % "4.1.1")
+
+addDependencyTreePlugin
```
Did you use any task besides `dependencyTree`? `dependencyTree` comes by default with sbt, no need to add `addDependencyTreePlugin`.

cheeseng replied:
@mkurz This is already removed, thanks.

cheeseng commented:
@mkurz I just pushed a quick commit according to your comments, do you mind to see if it is fit for merging now?

Thanks!

cheeseng commented:
@mkurz I have resolved the conflict and upgraded to use scalatest+selenium 3.2.12.1, which includes a new support for MS Edge browser, do you mind to review and consider merging this when it is convenient for you?

Thanks!

mkurz commented:
@cheeseng Sure I will, I was a bit busy the last days, I will be back working on Play soon. Stay tuned.

cheeseng commented:
@mkurz Ping!  I have resolved the conflict and this PR is ready for your review again.  Thanks!

mkurz commented:
Thanks  @cheeseng! I want to set up GitHub actions before merging this pull request (actually, this is the last repo in the Play organization missing GHA) to make sure this change passes all the tests.

cheeseng commented:
@mkurz That's a good idea, I am adding CI build in scalatest+selenium (one of the dependency here) and it seems to work: 

https://github.com/scalatest/scalatestplus-selenium/pull/20

But when I added the same ci.yml here it failed with this weird error: 

```
[error] Failed to derive version from git tags. Maybe run `git fetch --unshallow`? Version: 0.0.0+1-1e104898
```

I am not familiar with this, may be it is related to sbt-dynver?

cheeseng commented:
@mkurz By the way I think the current main is broken now?  When I tried to build it I got the following error: 

```
cheeseng@cheeseng-RAVEN:~/git/scalatestplus-play$ sbt clean test
[info] welcome to sbt 1.7.1 (GraalVM Community Java 1.8.0_302)
[info] loading global plugins from /home/cheeseng/.sbt/1.0/plugins
[info] loading settings for project scalatestplus-play-build from plugins.sbt ...
[info] loading project definition from /home/cheeseng/git/scalatestplus-play/project
/home/cheeseng/git/scalatestplus-play/build.sbt:35: warning: method sonatypeRepo in class ResolverFunctions is deprecated (since 1.7.0): Use sonatypeOssRepos instead
ThisBuild / resolvers += Resolver.sonatypeRepo("releases")
                                  ^
[info] loading settings for project scalatestplus-play-root from build.sbt ...
[info] set current project to scalatestplus-play-root (in build file:/home/cheeseng/git/scalatestplus-play/)
[success] Total time: 0 s, completed Jul 13, 2022 10:58:49 PM
[info] Updating 
[info] Resolved  dependencies
[warn] 
[warn] 	Note: Unresolved dependencies path:
[error] sbt.librarymanagement.ResolveException: Error downloading org.scalatestplus:selenium-4-1_2.13:3.1.4.0
[error]   Not found
[error]   Not found
[error]   not found: /home/cheeseng/.ivy2/local/org.scalatestplus/selenium-4-1_2.13/3.1.4.0/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/service/local/repositories/releases/content/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/snapshots/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://repo.artima.com/releases/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: /home/cheeseng/.m2/repository/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalactic-1457/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalatest-1753/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error] 	at lmcoursier.CoursierDependencyResolution.unresolvedWarningOrThrow(CoursierDependencyResolution.scala:345)
[error] 	at lmcoursier.CoursierDependencyResolution.$anonfun$update$38(CoursierDependencyResolution.scala:314)
[error] 	at scala.util.Either$LeftProjection.map(Either.scala:573)
[error] 	at lmcoursier.CoursierDependencyResolution.update(CoursierDependencyResolution.scala:314)
[error] 	at sbt.librarymanagement.DependencyResolution.update(DependencyResolution.scala:60)
[error] 	at sbt.internal.LibraryManagement$.resolve$1(LibraryManagement.scala:59)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$12(LibraryManagement.scala:133)
[error] 	at sbt.util.Tracked$.$anonfun$lastOutput$1(Tracked.scala:73)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$20(LibraryManagement.scala:146)
[error] 	at scala.util.control.Exception$Catch.apply(Exception.scala:228)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11(LibraryManagement.scala:146)
[error] 	at sbt.internal.LibraryManagement$.$anonfun$cachedUpdate$11$adapted(LibraryManagement.scala:127)
[error] 	at sbt.util.Tracked$.$anonfun$inputChangedW$1(Tracked.scala:219)
[error] 	at sbt.internal.LibraryManagement$.cachedUpdate(LibraryManagement.scala:160)
[error] 	at sbt.Classpaths$.$anonfun$updateTask0$1(Defaults.scala:3688)
[error] 	at scala.Function1.$anonfun$compose$1(Function1.scala:49)
[error] 	at sbt.internal.util.$tilde$greater.$anonfun$$u2219$1(TypeFunctions.scala:62)
[error] 	at sbt.std.Transform$$anon$4.work(Transform.scala:68)
[error] 	at sbt.Execute.$anonfun$submit$2(Execute.scala:282)
[error] 	at sbt.internal.util.ErrorHandling$.wideConvert(ErrorHandling.scala:23)
[error] 	at sbt.Execute.work(Execute.scala:291)
[error] 	at sbt.Execute.$anonfun$submit$1(Execute.scala:282)
[error] 	at sbt.ConcurrentRestrictions$$anon$4.$anonfun$submitValid$1(ConcurrentRestrictions.scala:265)
[error] 	at sbt.CompletionService$$anon$2.call(CompletionService.scala:64)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[error] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[error] 	at java.lang.Thread.run(Thread.java:748)
[error] (scalatestplus-play / update) sbt.librarymanagement.ResolveException: Error downloading org.scalatestplus:selenium-4-1_2.13:3.1.4.0
[error]   Not found
[error]   Not found
[error]   not found: /home/cheeseng/.ivy2/local/org.scalatestplus/selenium-4-1_2.13/3.1.4.0/ivys/ivy.xml
[error]   not found: https://repo1.maven.org/maven2/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/service/local/repositories/releases/content/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/snapshots/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://repo.artima.com/releases/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: /home/cheeseng/.m2/repository/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalactic-1457/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error]   not found: https://oss.sonatype.org/content/repositories/orgscalatest-1753/org/scalatestplus/selenium-4-1_2.13/3.1.4.0/selenium-4-1_2.13-3.1.4.0.pom
[error] Total time: 10 s, completed Jul 13, 2022 10:59:00 PM
```

cheeseng commented:
@mkurz I have added ci.yml that will run CI, here's the successfully run for this PR: 

https://github.com/playframework/scalatestplus-play/runs/7341113874?check_suite_focus=true

Cheers.

cheeseng commented:
@mkurz Ping again!  I have updated the PR to use the newly released ScalaTest 3.2.13, the CI tests run that included in this PR seems to be green also: 

https://github.com/playframework/scalatestplus-play/runs/7604177607?check_suite_focus=true

Cheers.

cheeseng commented:
@mkurz Ping again!  I have updated the PR to use selenium 4.4.0 and htmlunit 3.64.0, and the CI build is green.

mkurz commented:
@cheeseng Thanks again, fantastic! I will have a look asap :+1:

cheeseng commented:
@mkurz Ping again!  I fixed the conflict and added some fix to avoid flickering of test failures, the CI passed also, do you mind to look at this again?

Thanks.

ihostage commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,22 @@
+name: CI
```
Please, use Play [sharable workflow](https://github.com/playframework/.github) for CI 🙏 
See https://github.com/playframework/play-ws/blob/main/.github/workflows/build-test.yml as example 😉

cheeseng replied:
@ihostage Is there a way to use sharable workflow with macos image?  I think macos image can run best with our tests here with support of the most browser types.

ihostage replied:
Hmmm 🤔 What will we lose if we use Linux image?

cheeseng replied:
@ihostage We have selenium tests here that test for different types of browser, using the macos image we can run the tests for safari, chrome, firefox and msedge in addition to htmlunit, only the old internet explorer tests will be skipped.  Using linux I think we can only run chrome, firefox and msedge.

cheeseng replied:
@ihostage Perhaps I can keep the test job but try to use the sharable workflow for check-code-style, check-binary-compatibility, check-docs?

cheeseng replied:
@ihostage I studied the sharable workflow source a bit, the matrix approach looks nice, I wonder if it is possible to make the runs-on as an input with default value of 'ubuntu-20.04', but allow here in scalatest+play to override it as 'macos-latest' ?

cheeseng replied:
@ihostage Trying out the check-docs, but encountered this weird error: 

![Screenshot from 2022-09-21 12-00-40](https://user-images.githubusercontent.com/1322009/191411509-7be7f01c-4195-4bc4-91a1-a6235f64f975.png)

Mind to help seeing what's going wrong?  fyi I am trying out the sharable workflow in this PR: 

https://github.com/playframework/scalatestplus-play/pull/373

mkurz commented:
@cheeseng Sorry not being responsive here, I want to dedicate time tomorrow for this issue.
